File: sol/solution/0001-0100/0018/Solution.ex
defmodule Solution do
  @spec four_sum(nums :: [integer], target :: integer) :: [[integer]]
  def four_sum([], _target), do: []
  def four_sum([_], _target), do: []
  def four_sum([_, _], _target), do: []
  def four_sum([_, _, _], _target), do: []
  def four_sum([a, b, c, d], target) do
    if a + b + c + d == target do
      [[a, b, c, d]]
    else
      []
    end
  end
  def four_sum(nums, target) do
    nums =
      nums
      |> Enum.sort()
      |> List.to_tuple()
    len = tuple_size(nums)
    do_four_sum(nums, len, target, 0, [])
  end
  defp do_four_sum(nums, len, target, i, acc) do
    val_i = elem(nums, i)
    cond do
      (val_i > 0 and val_i > target) or i > len - 4 ->
        acc
      i == 0 or elem(nums, i - 1) != val_i ->
        triplets = do_three_sum(nums, len, target - val_i, i + 1, i + 1, [])
        acc = Enum.reduce(triplets, acc, fn triplet, acc -> [[val_i | triplet] | acc] end)
        do_four_sum(nums, len, target, i + 1, acc)
      true ->
        do_four_sum(nums, len, target, i + 1, acc)
    end
  end
  defp do_three_sum(nums, len, target, i, start_i, acc) do
    val_i = elem(nums, i)
    cond do
      (val_i > 0 and val_i > target) or i > len - 3 ->
        acc
      i == start_i or elem(nums, i - 1) != val_i ->
        acc = find_triplets(nums, target, i, i + 1, len - 1, acc)
        do_three_sum(nums, len, target, i + 1, start_i, acc)
      true ->
        do_three_sum(nums, len, target, i + 1, start_i, acc)
    end
  end
  defp find_triplets(_nums, _target, _i, l, r, acc) when l >= r, do: acc
  defp find_triplets(nums, target, i, l, r, acc) do
    val_i = elem(nums, i)
    val_l = elem(nums, l)
    val_r = elem(nums, r)
    sum = val_i + val_l + val_r
    cond do
      sum > target ->
        find_triplets(nums, target, i, l, r - 1, acc)
      sum < target ->
        find_triplets(nums, target, i, l + 1, r, acc)
      true ->
        acc = [[val_i, val_l, val_r] | acc]
        l = grow_from_left(nums, l + 1, r, val_l)
        r = shrink_from_right(nums, l, r - 1, val_r)
        find_triplets(nums, target, i, l, r, acc)
    end
  end
  defp shrink_from_right(nums, l, r, prev_val_r) do
    cond do
      r <= l or elem(nums, r) != prev_val_r -> r
      true -> shrink_from_right(nums, l, r - 1, prev_val_r)
    end
  end
  defp grow_from_left(nums, l, r, prev_val_l) do
    cond do
      l >= r or elem(nums, l) != prev_val_l -> l
      true -> grow_from_left(nums, l + 1, r, prev_val_l)
    end
  end
end

File: sol/solution/0001-0100/0018/Solution.erl
-spec four_sum(Nums :: [integer()], Target :: integer()) -> [[integer()]].
four_sum(Nums, Target) ->
    case length(Nums) < 4 of
        true -> [];
        false ->
            SortedNums = lists:sort(Nums),
            Array = array:from_list(SortedNums),
            N = length(SortedNums),
            four_sum_helper(Array, N, Target, 0, [])
    end.
four_sum_helper(Array, N, Target, I, Result) when I >= N - 3 -> Result;
four_sum_helper(Array, N, Target, I, Result) ->
    case I > 0 andalso array:get(I, Array) =:= array:get(I-1, Array) of
        true -> 
            four_sum_helper(Array, N, Target, I+1, Result);
        false ->
            NewResult = three_sum_helper(Array, N, Target, I, I+1, Result),
            four_sum_helper(Array, N, Target, I+1, NewResult)
    end.
three_sum_helper(Array, N, Target, I, J, Result) when J >= N - 2 -> Result;
three_sum_helper(Array, N, Target, I, J, Result) ->
    case J > I + 1 andalso array:get(J, Array) =:= array:get(J-1, Array) of
        true ->
            three_sum_helper(Array, N, Target, I, J+1, Result);
        false ->
            NewResult = two_pointer_helper(Array, N, Target, I, J, J+1, N-1, Result),
            three_sum_helper(Array, N, Target, I, J+1, NewResult)
    end.
two_pointer_helper(Array, _N, _Target, _I, _J, Left, Right, Result) when Left >= Right -> Result;
two_pointer_helper(Array, N, Target, I, J, Left, Right, Result) ->
    Sum = array:get(I, Array) + array:get(J, Array) + 
          array:get(Left, Array) + array:get(Right, Array),
    case Sum of
        Target ->
            Quadruplet = [array:get(I, Array), array:get(J, Array), 
                         array:get(Left, Array), array:get(Right, Array)],
            NewLeft = skip_duplicates_left(Array, Left, Right),
            NewRight = skip_duplicates_right(Array, Left, Right),
            two_pointer_helper(Array, N, Target, I, J, NewLeft, NewRight, [Quadruplet | Result]);
        Sum when Sum < Target ->
            two_pointer_helper(Array, N, Target, I, J, Left+1, Right, Result);
        _ ->
            two_pointer_helper(Array, N, Target, I, J, Left, Right-1, Result)
    end.
skip_duplicates_left(Array, Left, Right) ->
    skip_duplicates_left(Array, Left, Right, Left+1).
skip_duplicates_left(Array, Left, Right, NewLeft) when NewLeft < Right ->
    case array:get(NewLeft, Array) =:= array:get(Left, Array) of
        true -> skip_duplicates_left(Array, Left, Right, NewLeft+1);
        false -> NewLeft
    end;
skip_duplicates_left(_Array, _Left, Right, NewLeft) when NewLeft >= Right -> NewLeft.
skip_duplicates_right(Array, Left, Right) ->
    skip_duplicates_right(Array, Left, Right, Right-1).
skip_duplicates_right(Array, Left, Right, NewRight) when NewRight > Left ->
    case array:get(NewRight, Array) =:= array:get(Right, Array) of
        true -> skip_duplicates_right(Array, Left, Right, NewRight-1);
        false -> NewRight
    end;
skip_duplicates_right(_Array, Left, _Right, NewRight) when NewRight =< Left -> NewRight.
